### PR TITLE
[wasm][debugger] Fixing race condition

### LINKED ISSUE #64474: VS2022 Blazor debugging with F10/F11 goes into JS minimised code and not my code
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/VS2022-RC3-Blazor-debugging-with-F11-goe/1566868)._

---
[severity:It's more difficult to complete my work]
I'm calling a method and I want to step into that method with F10/F11 to continue debugging. However, I step into a big mess of JS Blazor minimised code. Not my thing.

See the recorded screen.

---
#### Feedback Bot on 10/28/2021, 11:38 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Thays Tagliaferri de Grazia [MSFT] on 11/3/2021, 05:40 AM: 

<p>I need help to reproduce it, I created a simple async function and could step into without any issue, could you please tell me if you only have this behavior when trying to step into this specific function GetUsers? If you remove everything that is in your watch window do you still have this behavior? Could you share the code of this GetUsers function?</p>
<p>Thanks a lot.</p>

#### Roberto Mencia Franco on 11/3/2021, 03:52 PM: 

<p>Sorry I can’t really replicate in a brand new solution, but I can replicate it in my solution consistently as shown in the recording.</p>
<p>You probably need to investigate internally in the source code to see the paths that could lead to the problem, as it happens all the time in my solution with F11.</p>

#### Roberto Mencia Franco on 11/7/2021, 09:42 PM: 

<p>I’ve bee getting this behaviour even with F10. It seems to stop in framework JS code when there are breakpoints in the lines of the code surrounding the area being debugged.</p>

#### Thays Tagliaferri de Grazia [MSFT] on 11/8/2021, 04:19 AM: 

<p>Could you please tell me the version of this package that is being used on Client and on Server project?</p>
<p>&lt;PackageReference Include=“Microsoft.AspNetCore.Components.WebAssembly.DevServer” Version=“VERSION” PrivateAssets=“all” /&gt;</p>

#### Roberto Mencia Franco on 11/8/2021, 02:11 PM: 

<p>The client:<br>
&lt;PackageReference Include=“Microsoft.AspNetCore.Components.WebAssembly.DevServer” Version=“6.0.0-rc.2.21480.10” PrivateAssets=“all” /&gt;</p>
<p>The server:<br>
&lt;PackageReference Include=“Microsoft.AspNetCore.Components.WebAssembly.Server” Version=“6.0.0-rc.2.21480.10” /&gt;</p>

#### Roberto Mencia Franco on 12/1/2021, 02:30 PM: 

<p>I have not heard from this issue. Debugging Blazor is a pain.<br>
I keep getting into JS code all the time, with F10 and F11. I can’t even debug a session.</p>
<p>I’m currently using the VS2022 17.1 pre1 and still happening.</p>
<p></p>

#### Thays Tagliaferri de Grazia [MSFT] on 12/6/2021, 08:15 AM: 

<p>We are still trying to reproduce it. Do you have any class library being used using debug type = full? Do you have anything that you can share with me about your project to help me to reproduce?</p>

#### Roberto Mencia Franco on 12/6/2021, 02:30 PM: 

<p>I don’t know how to consistently reproduce it, but it happens in pretty much every single debug session. And by far from all the large number of bugs that VS/Blazor has, this one is on the top for slowing down development.</p>
<p>Most of the times I work in code inside the “<strong>protected override async Task OnInitializedAsync()</strong>” of the pages.</p>
<p>I would probably say that it happens when calling async methods: <strong>var name = await c.GetAsync(…)</strong><br>
It happens with F10 and F11, but I would say that await async methods has a lot more lottery tickets.</p>
<p>This is an example in where it’s happened a few times yesterday.</p>
<p></p>

#### Thays Tagliaferri de Grazia [MSFT] on 12/7/2021, 08:27 AM: 

<p>When you start your Blazor WebAsssembly Application, pressing something like BlazorApp75.Server button, a console is opened, with a text like this:<br>
info: Microsoft.Hosting.Lifetime[14]<br>
Now listening on: <a target='_blank' href="https://localhost:5238">https://localhost:5238</a><br>
info: Microsoft.Hosting.Lifetime[0]<br>
Application started. Press Ctrl+C to shut down.<br>
info: Microsoft.Hosting.Lifetime[0]<br>
Hosting environment: Development<br>
info: Microsoft.Hosting.Lifetime[0]<br>
Content root path: T:\thays\source\repos\BlazorApp75\BlazorApp75\Server\</p>
<p>Can you share the content of this Console after you get the error when you press F10/F11?<br>
Also can you shared the content of the window in VS: Output -&gt; Show output from: Debug?</p>
<p>Thanks for your help.</p>

#### Roberto Mencia Franco on 12/7/2021, 03:03 PM: 

(private comment, text removed)
#### Thays Tagliaferri de Grazia [MSFT] on 12/8/2021, 06:32 AM: 

<p>Can you please run on IIS as you usually do but send me the value in the window:<br>
Output -&gt; then choose “Show output from: ISV.isEducation.Portal.Server - <a target='_blank' href="http://ASP.NET">ASP.NET</a> Core Web Server”.<br>
Please send me the output after you press F11 and you see the JS code.</p>

#### Feedback Bot on 12/14/2021, 09:38 PM: 

We will close this report in 14 days because we don’t have enough information to investigate further. To keep the problem open, please provide the requested details.

#### Roberto Mencia Franco on 12/15/2021, 01:20 PM: 

<p>I’m sorry that I have not been able to provide more information for, doing part of the work that the Testing team hasn’t been able to do.<br>
It won’t go away by automatically closing the bug. So, keep it open and me or someone will eventually do your job at troubleshooting the problem.<br>
In the meantime, I’ll keep paying my license.</p>

#### Roberto Mencia Franco on 12/16/2021, 08:12 PM: 

(private comment, text removed)
#### Roberto Mencia Franco on 12/16/2021, 08:17 PM: 

(private comment, text removed)
#### Feedback Bot on 1/18/2022, 07:54 AM: 

<p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by thaystg**
Unfortunatelly I wasn't able to reproduce it in a test case. To reproduce it using a Blazor app was also very difficult.

I created a separated buffer to receive the "pauses"(breakpoint, exception, user_break, etc.) from the debugger and another buffer to uses to the responses that we receive from debugger when we ask for some information (locals value, invoke method, object values, etc;)

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1427671
Fixes https://github.com/dotnet/runtime/issues/64474

Test case to reproduce:

Change the content of FetchData.razor in a Blazor app from the template:
```
@page "/fetchdata"
@using BlazorApp89.Shared
@using System.Collections.ObjectModel
@inject HttpClient Http

<PageTitle>Weather forecast</PageTitle>

<h1>Weather forecast</h1>

<p>This component demonstrates fetching data from the server.</p>

@if (forecasts == null)
{
    <p><em>Loading...</em></p>
}
else
{
    <table class="table">
        <thead>
            <tr>
                <th>Date</th>
                <th>Temp. (C)</th>
                <th>Temp. (F)</th>
                <th>Summary</th>
            </tr>
        </thead>
        <tbody>
            @foreach (var forecast in forecasts)
            {
                <tr>
                    <td>@forecast.Date.ToShortDateString()</td>
                    <td>@forecast.TemperatureC</td>
                    <td>@forecast.TemperatureF</td>
                    <td>@forecast.Summary</td>
                </tr>
            }
        </tbody>
    </table>
}

@code {
    private WeatherForecast[]? forecasts;

    protected override async Task OnInitializedAsync()
    {
        forecasts = await RetForecasts();
        var otherVar = await GetObservableCollection();
        Console.WriteLine(otherVar);
    }
    protected async Task<WeatherForecast[]> RetForecasts()
    {
        var a = await Http.GetFromJsonAsync<WeatherForecast[]>("WeatherForecast");
        var b = await Http.GetFromJsonAsync<WeatherForecast[]>("WeatherForecast");
        var c = await Http.GetFromJsonAsync<WeatherForecast[]>("WeatherForecast");
        var d = await Http.GetFromJsonAsync<WeatherForecast[]>("WeatherForecast");
        var e = await Http.GetFromJsonAsync<WeatherForecast[]>("WeatherForecast");
        Console.WriteLine(b);
        Console.WriteLine(c);
        Console.WriteLine(d);
        Console.WriteLine(e);
        return a;
    }
    protected async Task<ObservableCollection<WeatherForecast>> GetObservableCollection()
    {
        var forecastsLocal = await RetForecasts();
        return new ObservableCollection<WeatherForecast>(forecastsLocal.ToList<WeatherForecast>());
    }
}
```
Add a breakpoint on line: 46, step into hovering the mouse over the `WeatherForecast[]` on `RetForecasts` method.

### CODE REVIEW

File: src/mono/wasm/runtime/debug.ts [hunk 1]
```diff
@@ -46,6 +47,19 @@ export function mono_wasm_add_dbg_command_received(res_ok: boolean, id: number,
...
+    const assembly_data = new Uint8Array(Module.HEAPU8.buffer, buffer, buffer_len);
+    const base64String = toBase64StringImpl(assembly_data);
+    const buffer_obj = {
+        res_ok,
+        res: {
+            id,
+            value: base64String
+        }
+    };
+    commands_result = buffer_obj;
```
Comment (lewing) [inline src/mono/wasm/runtime/debug.ts:60]: did you consider doing something like:

```suggestion
    commands_result[id] = buffer_obj;
```
Then read clear the result after for that id in the caller?

**Review by lewing — COMMENTED** [review]
should we make the

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs [hunk 2]
```diff
@@ -700,7 +700,7 @@ public PointerValue(long address, int typeId, string varName)
     internal class MonoSDBHelper
     {
         private static int debuggerObjectId;
-        private static int cmdId;
+        private static int cmdId = 1;
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs:703]: `cmdId == 0` is a special one now. Can you add a comment for that here?

File: src/mono/wasm/runtime/debug.ts [hunk 3]
```diff
@@ -43,7 +43,7 @@ export function mono_wasm_add_dbg_command_received(res_ok: boolean, id: number,
             value: base64String
         }
     };
-    commands_received = buffer_obj;
+    commands_received.set(id, buffer_obj);
```
[the remark this answers is not in the document]
Comment (thaystg) [reply src/mono/wasm/runtime/debug.ts:46]: And if there is, should I print a warning?

File: src/mono/wasm/runtime/debug.ts [hunk 4]
```diff
@@ -63,7 +65,8 @@ export function mono_wasm_send_dbg_command_with_parms(id: number, command_set: n
     mono_wasm_malloc_and_set_debug_buffer(command_parameters);
     cwraps.mono_wasm_send_dbg_command_with_parms(id, command_set, command, _debugger_buffer, length, valtype, newvalue.toString());
 
-    const { res_ok, res } = commands_received;
+    const { res_ok, res } = commands_received.get(id) as CommandResponse;
+    commands_received.delete(id);
```
Comment (lewing) [inline src/mono/wasm/runtime/debug.ts:69]:
```suggestion
    const { res_ok, res } = commands_received.get(id) as CommandResponse;
    commands_received.delete(id);
```
I'd suggest making a method that combines get and delete

**Review by lewing — APPROVED** [review]
One small comment but looks good

**Review by lewing — APPROVED** [review]
👍

**Pull request merged**